File: app/src/main/java/com/faendir/lightning_launcher/multitool/util/LightningObjectFactory.java
package com.faendir.lightning_launcher.multitool.util;

import android.util.Log;
import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import com.faendir.lightning_launcher.multitool.MultiTool;
import com.faendir.lightning_launcher.multitool.proxy.JavaScript;
import com.faendir.lightning_launcher.multitool.proxy.Utils;

/**
 * @author F43nd1r
 * @since 07.11.2017
 */
@SuppressWarnings({"unused", "WeakerAccess"})
@Keep
public class LightningObjectFactory {
    private Utils utils;

    /**
     * Reflection constructor. Call init after calling this
     */
    public LightningObjectFactory() {
    }

    public LightningObjectFactory(Utils utils) {
        this.utils = utils;
    }

    public void init(EvalFunction eval, FunctionFactory asFunction) {
        this.utils = new Utils(eval, asFunction);
    }

    public JavaScript get(String className) {
        try {
            if (MultiTool.DEBUG) Log.d(MultiTool.LOG_TAG, "ObjectFactory loading " + className);
            //noinspection unchecked
            Class<? extends JavaScript> clazz = (Class<? extends JavaScript>) Class.forName(className);
            return clazz.getConstructor(Utils.class).newInstance(utils);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @FunctionalInterface
    public interface EvalFunction {
        Object eval(@NonNull String methodName, Object... parameters);
    }

    @FunctionalInterface
    public interface FunctionFactory {
        Object asFunction(@NonNull Object target);
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/util/LightningObjectFactory.kt
package com.faendir.lightning_launcher.multitool.util

import android.util.Log
import androidx.annotation.Keep
import com.faendir.lightning_launcher.multitool.MultiTool
import com.faendir.lightning_launcher.multitool.proxy.JavaScript
import com.faendir.lightning_launcher.multitool.proxy.Utils

/**
 * @author F43nd1r
 * @since 07.11.2017
 */
@Keep
class LightningObjectFactory {
    private var utils: Utils? = null

    /**
     * Reflection constructor. Call init after calling this
     */
    @Suppress("unused")
    constructor()

    constructor(utils: Utils) {
        this.utils = utils
    }

    fun init(eval: (String, Array<Any?>)->Unit, asFunction: (Any) -> Any) {
        this.utils = Utils(eval, asFunction)
    }

    operator fun get(className: String): JavaScript {
        try {
            if (MultiTool.DEBUG) Log.d(MultiTool.LOG_TAG, "ObjectFactory loading $className")
            @Suppress("UNCHECKED_CAST")
            val clazz = Class.forName(className) as Class<out JavaScript>
            return clazz.getConstructor(Utils::class.java).newInstance(utils)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }

    @FunctionalInterface
    interface FunctionFactory {
        fun asFunction(target: Any): Any
    }
}
